cvat/apps/engine/media_extractors.py

```diff
@@ -147,92 +150,84 @@ def __init__(self, source_path, step=1, start=0, stop=0):
 
-    def __iter__(self):
-        for f in zip(self._source_path, self.image_names):
-            yield (BytesIO(self._zip_source.read(f[0])), f[1])
-
-    def __len__(self):
-        return len(self._source_path)
-
-    def __getitem__(self, k):
-        return (BytesIO(self._zip_source.read(self._source_path[k])), self.image_names[k])
-
     def __del__(self):
         self._zip_source.close()
 
-    def save_preview(self, preview_path):
-        with open(preview_path, 'wb') as f:
-            f.write(self._zip_source.read(self._source_path[0]))
+    def get_preview(self):
+        io_image = io.BytesIO(self._zip_source.read(self._source_path[0]))
+        return self._get_preview(io_image)
+
+    def get_image(self, i):
+        return io.BytesIO(self._zip_source.read(self._source_path[i]))
 
-    @property
-    def image_names(self):
-        return [os.path.join(os.path.dirname(self._zip_source.filename), p) for p in self._source_path]
+    def get_path(self, i):
+        return os.path.join(os.path.dirname(self._zip_source.filename), self._source_path[i])
 
 class VideoReader(IMediaReader):
     def __init__(self, source_path, step=1, start=0, stop=0):
-        self._output_fps = 25
-
         super().__init__(
             source_path=source_path,
             step=step,
             start=start,
             stop=stop,
         )
 
-    def __iter__(self):
-        def decode_frames(container):
-            for packet in container.demux():
-                if packet.stream.type == 'video':
-                    for frame in packet.decode():
-                        yield frame
+    def _has_frame(self, i):
+        if i >= self._start:
+            if (i - self._start) % self._step == 0:
+                if self._stop == 0 or i <= self._stop:
+                    return True
+
+        return False
+
+    def _decode(self, container):
+        frame_num = 0
+        for packet in container.demux():
+            if packet.stream.type == 'video':
+                for image in packet.decode():
+                    frame_num += 1
+                    if self._has_frame(frame_num - 1):
+                        yield (image, self._source_path[0], image.pts)
 
+    def __iter__(self):
         container = self._get_av_container()
         source_video_stream = container.streams.video[0]
         source_video_stream.thread_type = 'AUTO'
-        image_names = self.image_names
 
-        return itertools.zip_longest(decode_frames(container), image_names, fillvalue=image_names[0])
+        return self._decode(container)
 
-    def __len__(self):
+    def get_progress(self, pos):
         container = self._get_av_container()
         # Not for all containers return real value
-        length = container.streams.video[0].frames
-        return length
-
-    def __getitem__(self, k):
-        return next(itertools.islice(self, k, k + 1))
+        stream = container.streams.video[0]
+        return pos / stream.duration
```

Comment:
I think it may be the same problem as for stream.frames, one of solutions: show number of extracted frames for status in this case. What do you think?